### ISSUE TITLE
Save on client only

### ISSUE THREAD

**Issue by s7dhansh** [jagi/meteor-astronomy#274]
@jagi is there a feature where we can temporarily update only on client, say via `.save({clientOnly: true})` ?

If not, would you be willing to implement such a feature in 2.0? It will help a lot with form validations/dynamic displays, where let's say we have to show/hide some sections based on certain fields. Currently, we need to use Session/Reactive variables, but with such an option, the client data will change, triggering the form changes too.

We can also do client side validations with a similar concept `save({clientOnly: true, validate: true})`.
In the end, we can save it at the server via a simple `save(/*{validate: true}*/)` (or a Meteor.call depending on the security level we want to maintain).

**Comment by jagi** [comment]
It's not possible right now and it won't probably be possible in 2.0. I know that using ReactiveVar may not be the most convenient way of doing it. The problem with saving a document only on the client is that data may be overridden by a change on the server. So it won't solve your problem. You will till need to use ReactiveVar or similar.

**Comment by s7dhansh** [comment]
Hmm.. ok. Later sometime then, hopefully :)

**Comment by jagi** [comment]
It would be more handy to make astronomy documents reactive.

```js
var user = new User();
// use user in the form
user.set('firstName', 'John'); /* or */ user.firstName = John; // Would reactively update the form.
```
We could achieve reactivity by introducing some extra module for Astronomy 2.0. It would have to use setters/getters to allow that but it's possible. I will investigate this problem.

**Comment by s7dhansh** [comment]
Yes, that sounds better!

**Comment by jagi** [comment]
Probably I will also implement saving on the client only as I may need it too. So it wasn't bad idea. More information soon

**Comment by daveeel** [comment]
Any update on this? I'm recently running into client side form handing, the data sets are not to be saved anywhere but throw to REST endpoints to consume after checking. So thinking to use Astronomy for client side only validation etc...

**Comment by lukejagodzinski** [comment]
After thinking about this I'm not so sure if it should be implemented. However, you can still use client side only collection.

``` js
// on the client
const Posts = new Mongo.Collection(null); // no collection name, client side only
const Post = Class.create({
  name: 'Post',
  collection: Posts,
  fields: { /* ... */ }
});
```

**Comment by s7dhansh** [comment]
Why do you say so? Why should it not be implemented? Can you please
elaborate?

**Comment by daveeel** [comment]
I tried using null collection on client side. However something like

record.save()

won't work as expected. The local record change is not updated. Any clue?

**Comment by s7dhansh** [comment]
@daveeel can you give the sample code, or preferably a small reproduction

**Comment by daveeel** [comment]
@s7dhansh , shall make an isolated reproduction and report back here.

**Comment by lukejagodzinski** [comment]
@s7dhansh because if collection is defined for both client and server then what's the point of saving it only on the client, if there are client only collections (`new Mongo.Collection(null)`). I don't see any benefit of having such feature. It only introduces more complex API in Astronomy

**Comment by daveeel** [comment]
@jagi maybe I need your help to confirm the expected behaviour of client side 'null' collection:

Say I have a collection declared by:

```
@Entities = new Mongo.Collection null
@Entity = Astro.Class do
  name: 'Entity'
  collection: Entities
  fields:
    gender:
      type: 'string'
```

For a record say rec. I could do the following which is fine:

```
rec.gender = 'f'
```

What I wish is to be able to do is

```
rec.save()
```

So that I could have updated records by `Entities.find()`. Which is what I mean by client side save.

Now `rec.save()` has no effect on the client side only collection. Is that the expected behaviour ?

**Comment by s7dhansh** [comment]
@jagi Let me give you an example. Let's say there are 2 templates T1 and T2. Based on a value change through a form in T1, I want to do something in T2. Currently, it would be possible through Session (or states in react). If astronomy allows such a client side only save, it will ensure that Collection.findOne() will give the current client side value which can be used for any further interactions.

T1's form would get saved server side too, but only later (to decrease the number of server calls). So it does not make sense to use a client side collection. In complex apps, such a scenario is quite common, which is why I requested for this feature.

**Comment by s7dhansh** [comment]
@daveeel Are you using 2.0? Does a manual mongo rec.update() work on the client side for you in your code?

**Comment by lukejagodzinski** [comment]
@s7dhansh you can put a document in a reactive variable and modify it freely and use it to display form fields. You don't have to save it anywhere. That way, you can postpone save operation.

**Comment by daveeel** [comment]
@s7dhansh , I used 1.2.  And also tried 2.0 after reading your comment. Mongo update() does work and it's my original way to make changes. I'm trying to use Astronomy for centralised client side data validation. It seems that I have to hook server side anyway.

By the way, your use case is client side state management. Session is the old way. Reactive-Var is also clumsy for inter-template communication. Have you tried ViewModel ? ( https://viewmodel.org ). https://viewmodel.org/docs/viewmodels#share is what you need.

**Comment by s7dhansh** [comment]
@daveeel In 1.2, you need to use `rec.set("gender", "f")` and not `rec.gender = 'f'`.

Thank you. Viewmodel does look nice. But now I am shifting to React slowly. Is there any similar package for React?

Such a feature will be quite suitable and good for astronomy, I think. Will try to implement it myself sometime.

**Comment by daveeel** [comment]
@s7dhansh , thanks for quick reply. Shall try what you suggested again in 1.2. In 2.0, both makes no difference for null client collection. I tried with one record only. Say:

```
r = Entity.findOne() # where .gender == ''
r.gender = 'f' # also tried r.set('gender','f')
r.save() # no use, but I just tried
```

Now I tried both `Entity.findOne()` and `Entities.findOne()` and hope to find the 'updated' record, but nothing changed ...

Anyways, will switch to server collection.

I'm not a react convert yet, so not much input about React. But seems that most go for Redux for client side state management, including MDG. FYI, the ViewModel author is working on a React based VM.
